### ISSUE TITLE
Set styled object as a whole based on props

### ISSUE THREAD

**Issue by RickEyre** [cssinjs/styled-jss#41]
Is it possible / would you accept a pull request that could accomplish something of the following:

```js
const CaseOne = {
  color: red
};

const CaseTwo = {
  color: blue
};

const MyStyledComponent = styled('button')(
  props => props.caseOne ? CaseOne : CaseTwo
);
```

I know currently you can set CSS properties individually based on props, but the library doesn't support replacing the whole styled object based on props. 

This is a trivial example case, but I've had a number of use cases using this library where the styled objects were so different that I'd rather just swap them in and out as a whole.

I've also had a few cases where the styled objects are coming from different node modules and it's not sensible to override them one by one; I also don't want to know what the styled object looks like because it's coming from a different node module.

**Comment by lttb** [comment]
hi @RickEyre 

Thanks for this suggestion! 

But I see some problems with this feature, and the most meaningful problem for me is possible performance downgrade for this case. So let me explain how `styled-jss` works now.

When we **mount** styled component created with `styled(tag)(styles)` **for the first time** ,`jss` parses `styles` object and creates the `Rule` from it, that will be used in CSSOM. So in this process`jss` also handles [function values](https://github.com/cssinjs/jss/blob/master/docs/json-api.md#function-values), applies different `jss-plugins` and so on. **And it's quite expensive at all**. But we've got many optimizations later, when we reuse or update this component, like:
- `styled-jss` reuse this calculations, and it does not parse `styles` again
- [function values](https://github.com/cssinjs/jss/blob/master/docs/json-api.md#function-values) are [really fast on update](https://medium.com/dailyjs/high-performance-dynamic-styles-db28c873940)

That's why `styled-jss` is [one of the fastest `cssinjs` solutions](https://github.com/A-gambit/CSS-IN-JS-Benchmarks/blob/c018fb1cfadd28a0da9768127fc084f78ad7862a/RESULT.md) without inline-styles on rerender. And of course we're trying to improve our `first-mount time` too.

Let's take a look on this example:

```jsx
const cases = {
  one: {color: 'red'},
  two: {color: 'glue'}
}

const StyledComponent = styled('button')({
  color: props => cases[props.type]
})

// ...

export default (props) => (
  <div>
    <StyledComponent type="one">button 1</StyledComponent>
    <StyledComponent type="two">button 2</StyledComponent>
    <StyledComponent type={props.something ? 'one' : 'two'}>button</StyledComponent>
  </div>
)   
```

So on each rerender and mount we will just recalculate the `dynamic value` of created `StyledComponent Rule` with function `color: props => cases[props.type]` for each instance, and it will be cheap and fast.

And now let's check an example with suggested behavior:

```jsx
const CaseOne = {
  color: red
};

const CaseTwo = {
  color: blue
};

const MyStyledComponent = styled('button')(
  props => props.caseOne ? CaseOne : CaseTwo
);
```

To make this thing work, we need to recalculate the whole styles object for each render and mount, that means that we should **parse styles object and fully update or create the Rule every time again**. It seems that it may cause some performance problems.

We will try to improve our `first-mount` time and make `styled-jss` faster at all, but nowadays, unfortunately, I don't see the common effective solution for such behavior.

May be you have some suggestions?

**Comment by zerobias** [comment]
Oops, noted that it still doesn't solve issue yet, so I hope at least it yields some good thoughts

**Comment by goodmind** [comment]
@lttb does `glamorous` also do recalculation for each render?

**Comment by kof** [comment]
function rules have been added to the recent jss version, so this should work out of the box or easy to add now

**Comment by goodmind** [comment]
@lttb do you plan to add it?

**Comment by lttb** [comment]
@goodmind yeah, I'm almost done with it here https://github.com/cssinjs/styled-jss/pull/35

There are some updates to API:
https://github.com/cssinjs/styled-jss/blob/2dd5de959c782ef23c01d1780632a3a336342e66/README.md#theming

**Comment by hendrul** [comment]
What if I want to use more than one function rules?, In my case I want to use the utility functions offered by [styled-system](https://github.com/jxnblk/styled-system) which would work perfectly if you support this case

**Comment by lttb** [comment]
Resolved in https://github.com/cssinjs/styled-jss/pull/35

**Comment by lttb** [comment]
Released in 2.2.0.

You can check the example here: https://codesandbox.io/s/y0162p38lv
